cli - Modify cockroach demo to allow for add node

ajstorm opened the pull request:
Commit `cea1bfe`: cli - Modify cockroach demo to allow for add node

Previously, cockroach demo allowed you to start a 9 node cluster, but
once activated, it was not possible to add nodes to the cluster.  This
commits adds the ability to add nodes to a running demo cluster.

The main motivation for this change is to more easily test multi-region
clusters.

Release note (cli change): `cockroach demo` now allows for nodes to be
added using the `\demo` client-side command.  This works in both single
node and multi-node configurations (e.g. when started with `--nodes int`
or `--geo-partitioned-replicas`).

---

Commit `8dd0f48`: sql: Change SHOW REGIONS to pull from kv_node_status table

Previously the SHOW REGIONS statement pulled data from the
`crdb_internal.gossip_nodes` table.  The problem with this table however,
is that it isn't a consistent and reliable source of information.  For
example, nodes remain in the gossip table for 24 hours after they've
been shutdown and decommissioned.  Additionally, nodes may not be
present in the `gossip_nodes` table immediately after they've been added
to the cluster.  To fix these issues, SHOW REGIONS will now pull data
from the `crdb_internal.kv_node_status` table, to which nodes are added when they are
first started, and removed when they are decommissioned.

This commit fixes #56332.

Release note (sql change): SHOW REGIONS has changed the column name for
availability zones to `zones` from `availability_zones`.  This was
agreed upon with the TLs and PMs for multi-region simplification.

ajstorm committed:
cli - Modify cockroach demo to allow for add node

Previously, cockroach demo allowed you to start a 9 node cluster, but
once activated, it was not possible to add nodes to the cluster.  This
commits adds the ability to add nodes to a running demo cluster.

The main motivation for this change is to more easily test multi-region
clusters.

Release note (cli change): `cockroach demo` now allows for nodes to be
added using the `\demo` client-side command.  This works in both single
node and multi-node configurations (e.g. when started with `--nodes int`
or `--geo-partitioned-replicas`).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56344)
<!-- Reviewable:end -->

otan commented on pkg/cli/sql.go:
```diff
@@ -540,6 +546,22 @@ func (c *cliState) handleDemo(cmd []string, nextState, errState cliStateEnum) cl
 	}
 
+	// Special case the add command it takes a string instead of a node number
+	if cmd[0] == "add" {
```
optional personal nit: would prefer all the cmd[0]s to be switched in the same "area", maybe something like:

```
switch cmd[0] {
   case "add":
     ....
   case "recommission", "decommission", ...:
     <handle parse node id, etc, or maybe put this entire section into a new function
     switch cmd[0] {
        case "recommission": // ....
      }
}

```

so we're upfront about all possible states and cmd[0] is in one place. also i anticipate `add` taking more arguments in future (i imagined at a later stage we might want this such that `\demo add` would take the same flags as `cockroach start`) and the len(cmd) != 2 check may need to be moved lower.

otan reviewed (commented):
looking good, not on cli-prs so maybe @knz also wants a squiz

knz reviewed (commented):
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/cli/demo_cluster.go, line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMZ-r1-503:-MLTf3KcEN1IZ9Tbwp1C:b-e01k9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L503)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: use ``` `"'` ``` to avoid the escaping.
</blockquote></details>

May I suggest a regexp here `\"[...]+\"|'[...]+'|[...]+` so as to ensure the quotes are balanced, and no special character is present inside the locality identifier.
___
*[pkg/cli/demo_cluster.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTf_3M9c2lQG_L_-gJ:-MLTf_3M9c2lQG_L_-gK:b-t6jhu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L505)):*
> ```Go
> 	trimedString := strings.Trim(localityString, "\"'")
> 
> 	// Setup locality based on supplied string
> ```

nit, here and throughout the diff: comments are sentences and thus end with a period.
___
*[pkg/cli/sql.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfjAU4d2dYXvamNcd:-MLTfjAU4d2dYXvamNce:bf2l446) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L101)):*
> ```Go
>   \demo decommission <nodeid>  decommission a node.
>   \demo recommission <nodeid>  recommission a node.
>   \demo add <locality>         add a node (locality specified as "region=<region>,zone=<zone>").
> ```

it's unclear what localities are valid. May I suggest another client command alongside to list available localities already defined in the cluster + suggestions.
___
*[pkg/cli/sql.go, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfgAf5wfIvMqu7MRw:-MLTfgAf5wfIvMqu7MRx:b-tmkhzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L536)):*
> ```Go
> 
> 	// The \demo command has one of three patterns:
> 	// 	- A lone command (currently, only ls)
> ```

nit: spacing
___
*[pkg/cli/sql.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-559:-MLTfx5a0TxcRcNlx8xe:b-b8fobr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L559)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: use %q instead of \"%s\" (they do the same thing)
</blockquote></details>

(pedantic: `%q` does a few more things than `"%s"` if the string contains special characters. But here they are equivalent, yes, because localities do not contain such special characters.)
___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLTgFQu3uwd_zhZ3pBA:b-69u0x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
> ```Go
> func (d *delegator) delegateShowRegions(n *tree.ShowRegions) (tree.Statement, error) {
> 	sqltelemetry.IncrementShowCounter(sqltelemetry.Regions)
> 	// FIXME (storm): Change this so that (a) it doesn't use hard-coded strings
> ```

1) FYI we don't leave "FIXME" in PRs when they are merged. I am assuming you are retaining this comment here because the PR is not complete. When it is, and if you think there is still work remaining to be done, use "TODO" instead.

2) I agree this code is incomplete as-is. I think the way you wrote this is brittle. I am not even sure we have any standard way to identify parts of the locality hierarchy throughout cockroachdb.

In my understanding, this hierarchy with the names "regions", "az", "availability-zone" and "zone" are colloquialisms in the current state of our docs. We can assume our users may have a different view on this. I would therefore refrain from adding a new SHOW command which crystallizes this choice in the code -- we have zero evidence it would work well for all users.

I appreciate you may need a custom solution for `cockroach demo`. If that is the case, I recommend you keep this solution in the `demo` code client-side, until we define a more standard structure in our zone configs server-side..


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
@otan @knz as this is really my first PR which does more than a trivial amount of golang/crdb coding, thanks for your patience.  All the comments have been resolved below (except for the one about detecting special characters in the locality flag.  Please see my thoughts about that below.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-:-MLhKtALDsC4_jYlIKZe:bqrq6yg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)

___
*[pkg/cli/demo_cluster.go, line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMZ-r1-503:-MLUBj916FnB6yWJ66X_:bcuogeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L503)):*
> \"[...]+\"|'[...]+'|[...]+
The balancing makes sense to me, but I'm not sure about the guarding for special characters.  We don't do that for the locality flag for cockroach start, so it would seem odd to me to do it here.  Also, while we could do it in both places (here and cockroach start), that could break customers who have special characters in their strings today (to say nothing of the fact that we'd expect  '=' to be in there).  I've made the change to do the balancing, and Oliver's suggestion to avoid the escaping.  Let me know what you think.
___
*[pkg/cli/demo_cluster.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTf_3M9c2lQG_L_-gJ:-MLgyAuc5vcnAYlHPmE9:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L505)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit, here and throughout the diff: comments are sentences and thus end with a period.
</blockquote></details>

Fixed
___
*[pkg/cli/sql.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfjAU4d2dYXvamNcd:-MLgyJS30Ak9iEUA2tNP:bk2vys2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

it's unclear what localities are valid. May I suggest another client command alongside to list available localities already defined in the cluster + suggestions.
</blockquote></details>

That command was recently added.  It's called SHOW REGIONS.
___
*[pkg/cli/sql.go, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfgAf5wfIvMqu7MRw:-MLgyZF14LowHyVK6vQW:b8wbqa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L536)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: spacing
</blockquote></details>

This was intentional to create a list which would stand out.  Is there a better way to do this?  With tabs?
___
*[pkg/cli/sql.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-550:-MLgyzu06fR3lmcgZsto:b-8alne6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L550)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

optional personal nit: would prefer all the cmd[0]s to be switched in the same "area", maybe something like:

```
switch cmd[0] {
   case "add":
     ....
   case "recommission", "decommission", ...:
     <handle parse node id, etc, or maybe put this entire section into a new function
     switch cmd[0] {
        case "recommission": // ....
      }
}

```

so we're upfront about all possible states and cmd[0] is in one place. also i anticipate `add` taking more arguments in future (i imagined at a later stage we might want this such that `\demo add` would take the same flags as `cockroach start`) and the len(cmd) != 2 check may need to be moved lower. 
</blockquote></details>

I had that at one point but the nested case statement to handle the commands which took a numeric vs a string struck me as looking more confusing than what I finally landed on.  If you and @knz think that the above looks better, I'm happy to make that change.
___
*[pkg/cli/sql.go, line 552 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-552:-MLgzRtXENjDS_L4NwBe:bdpvpby) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L552)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

should this be a TODO or nah?
</blockquote></details>

My plan was never to support it, as it would be tricky to specify new latency pairs for this newly added node and all other existing nodes.  
___
*[pkg/cli/sql.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-559:-MLgzjw58RwxaB5Uu9SS:by3u4o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L559)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

(pedantic: `%q` does a few more things than `"%s"` if the string contains special characters. But here they are equivalent, yes, because localities do not contain such special characters.)
</blockquote></details>

Actually, I think they will contain special characters (notably '=').  From the documentation, it seems as though %q requires the string to be double quoted though, which may not be the case here.  I think that implies that I need to stick with %s.
___
*[pkg/cli/interactive_tests/test_demo_node_cmds.tcl, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMa-r1-96:-MLhJngp4YUtKNcNjyOU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/interactive_tests/test_demo_node_cmds.tcl#L96)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can you try shutting down the node and check that still works as expected?
</blockquote></details>

done
___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMc-r1-21:-MLhJtTx52tsOAAjDhtg:b-6ab9za) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

missing (b)?
</blockquote></details>

:) There once was a (b) and a (c), but I addressed both of those and didn't remove the (a).  It's gone now.
___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLhKNpc2oYkgrlN3VGF:b-qfv4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) FYI we don't leave "FIXME" in PRs when they are merged. I am assuming you are retaining this comment here because the PR is not complete. When it is, and if you think there is still work remaining to be done, use "TODO" instead.

2) I agree this code is incomplete as-is. I think the way you wrote this is brittle. I am not even sure we have any standard way to identify parts of the locality hierarchy throughout cockroachdb.

In my understanding, this hierarchy with the names "regions", "az", "availability-zone" and "zone" are colloquialisms in the current state of our docs. We can assume our users may have a different view on this. I would therefore refrain from adding a new SHOW command which crystallizes this choice in the code -- we have zero evidence it would work well for all users.

I appreciate you may need a custom solution for `cockroach demo`. If that is the case, I recommend you keep this solution in the `demo` code client-side, until we define a more standard structure in our zone configs server-side..
</blockquote></details>

Agree on all accounts.  My next PR will clean this one up to address your comments.  Was planning on doing it with this PR, but then I realized that it seemed like a PR in itself (as we'll likely want to do some work on the locality flag as passed to cockroach start).  I've added a TODO to reflect this fact.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/cli/demo_cluster.go, line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMZ-r1-503:-MLjLNU5E_tS4KS90aLR:b-grobwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L503)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

> \"[...]+\"|'[...]+'|[...]+
The balancing makes sense to me, but I'm not sure about the guarding for special characters.  We don't do that for the locality flag for cockroach start, so it would seem odd to me to do it here.  Also, while we could do it in both places (here and cockroach start), that could break customers who have special characters in their strings today (to say nothing of the fact that we'd expect  '=' to be in there).  I've made the change to do the balancing, and Oliver's suggestion to avoid the escaping.  Let me know what you think.
</blockquote></details>

still looks like we can use ``` `"'` ``` to do escaping here?
also i think the past tense of trimed is trimmed
___
*[pkg/cli/sql.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-550:-MLjLfa8CN9vwZfdmUP1:blm8a3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L550)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I had that at one point but the nested case statement to handle the commands which took a numeric vs a string struck me as looking more confusing than what I finally landed on.  If you and @knz think that the above looks better, I'm happy to make that change.
</blockquote></details>

i think it looks neater, so long as the nested switch is inside a new function ;P 
___
*[pkg/cli/sql.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-559:-MLjLnql8_GSfdNgbIvY:by71qaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L559)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Actually, I think they will contain special characters (notably '=').  From the documentation, it seems as though %q requires the string to be double quoted though, which may not be the case here.  I think that implies that I need to stick with %s.
</blockquote></details>

in that case, we can use ``` `node %v ... "%s"` ``` to avoid the escaping.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/cli/demo_cluster.go, line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMZ-r1-503:-MLjZ2_m1JDzGXNLPCEw:b-xspjmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/demo_cluster.go#L503)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

still looks like we can use ``` `"'` ``` to do escaping here?
also i think the past tense of trimed is trimmed
</blockquote></details>

Fixed both.  I got a bit carried away with the trimming, and trimmed too many m's.
___
*[pkg/cli/sql.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-550:-MLjcp-q4ZnoE_9OcabS:b2qqm0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L550)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think it looks neater, so long as the nested switch is inside a new function ;P 
</blockquote></details>

I'm going to yield to @knz here as the tie-breaker.  Rafael, do you think this warrants a rewrite?
___
*[pkg/cli/sql.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-559:-MLjb1Cu7vvvjUnGYGfa:b73id9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L559)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

in that case, we can use ``` `node %v ... "%s"` ``` to avoid the escaping.
</blockquote></details>

As discussed over Slack, it's tricky to use backticks here, because of the CR.  Oliver agreed that we should leave it as-is.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @dt, @knz, and @otan)

___
*[pkg/cli/demo_cluster.go, line 505 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmPxwk3M7EOiU2HlXi:-MLmPxwk3M7EOiU2HlXj:bz1j3xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/demo_cluster.go#L505)):*
> ```Go
> 	// quoted strings, strip off the quotes.  Before we do that though, make sure that the quotes match,
> 	// or that there aren't any quotes in the string.
> 	re := regexp.MustCompile(`".+"|'.+'|[^'"]+`)
> ```

maybe turn that `.` into `[^"']` or `[-a-zA-Z0-9]` or whatever matches what we accept for localities.
___
*[pkg/cli/sql.go, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfgAf5wfIvMqu7MRw:-MLmQ_D-2Lt4vuXcRqzm:b82730s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L536)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

This was intentional to create a list which would stand out.  Is there a better way to do this?  With tabs?
</blockquote></details>

an idea: make a newline before the intro sentence at the top and the first hyphen,
then another newline before the list.

(my comment was specific about the fact the number of spaces before the first hyphen was not the same as on the subsequent lines)
___
*[pkg/cli/sql.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-550:-MLmQkjj2PlDqu_b6L5Z:bc6bdlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L550)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I'm going to yield to @knz here as the tie-breaker.  Rafael, do you think this warrants a rewrite?
</blockquote></details>

I don't care too much about the conditional, but I concur that this function body is now long enough it could be split into multiple functions.
___
*[pkg/cli/sql.go, line 539 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmQYvy7nLfPGC3tr7y:-MLmQYvy7nLfPGC3tr7z:b-hf44mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/sql.go#L539)):*
> ```Go
> 	//	- A command followed by a string (add followed by locality string)
> 	//	- A command followed by a node number (shutdown, restart, decommission, recommission)
> 	// We parse these commands separately, in the following blocks
> ```

nit: here and below, punctuation
___
*[pkg/cli/sql.go, line 549 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmQpvy9HThoFmafc_R:-MLmQpvy9HThoFmafc_S:b5c06bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/sql.go#L549)):*
> ```Go
> 	}
> 
> 	// Special case the add command it takes a string instead of a node number
> ```

nit: punctuation
___
*[pkg/cli/sql.go, line 564 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmQquf0o3gvl6scKwF:-MLmQqugEDUmhSdq6YmO:b5c06bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/sql.go#L564)):*
> ```Go
> 	}
> 
> 	// Begin handling the list of commands which take a node number
> ```

nit: punctuation
___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLmQB97A0TXDNN4Al5f:b3tz7zl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Agree on all accounts.  My next PR will clean this one up to address your comments.  Was planning on doing it with this PR, but then I realized that it seemed like a PR in itself (as we'll likely want to do some work on the locality flag as passed to cockroach start).  I've added a TODO to reflect this fact.
</blockquote></details>

As I wrote above I am not keen to introduce a new user-facing feature with committed support (SHOW) in a way that's not guaranteed to work for 100% of users.

You can either turn your SHOW statement into a new `crdb_internal` view (at the very least) or keep it in `demo.go` altogether.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/cli/demo_cluster.go, line 505 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmPxwk3M7EOiU2HlXi:-MLm_H-6DsXm7RzPCIaH:bcxs758) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/demo_cluster.go#L505)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe turn that `.` into `[^"']` or `[-a-zA-Z0-9]` or whatever matches what we accept for localities.
</blockquote></details>

I might be missing something, but I don't see how [^"'] would help here.  In both cases, once you hit the first ' or " you're going to terminate the regex search and if there's still more string left, the check below is going to fail.  

I'm reluctant to implement the second option as well, since we do locality string validity checking further down inside of loc.Set().  This will validate that the locality is of the form "key=value,key=value,..." (see test case added with "blah").  My preference is to not go overboard with validation here, especially since we're not doing much in the way of validation on cockroach start (a separate problem, but one that's not easy to fix without the potential to break existing customers.
___
*[pkg/cli/sql.go, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTfgAf5wfIvMqu7MRw:-MLmb1zP9_F2ZBtCo40P:bn3ug9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L536)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

an idea: make a newline before the intro sentence at the top and the first hyphen,
then another newline before the list.

(my comment was specific about the fact the number of spaces before the first hyphen was not the same as on the subsequent lines)
</blockquote></details>

Ack.  I'll blame this one on my goland setup which shows tabs and spaces in a weird way.  It's fixed now.  Added the desired new line too.
___
*[pkg/cli/sql.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLPZisQS6095-abYJMb-r1-550:-MLmvrTpEWsK5kcrt_B_:bc82ssc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/cli/sql.go#L550)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't care too much about the conditional, but I concur that this function body is now long enough it could be split into multiple functions.
</blockquote></details>

I've split it into multiple function and agree that it reads better this way.  Thanks for the encouragement.
___
*[pkg/cli/sql.go, line 539 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLmQYvy7nLfPGC3tr7y:-MLmbQ3Y7AM4sXl81lvl:bwovhlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ead961366eafac36ef97e9a406504a91d2286385/pkg/cli/sql.go#L539)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: here and below, punctuation
</blockquote></details>

Fixed.  This one will take me some time to get used to as in my old role we didn't consider comments to be sentences.
___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLmd20r0nf2hnORBZCb:b-y5vv7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

As I wrote above I am not keen to introduce a new user-facing feature with committed support (SHOW) in a way that's not guaranteed to work for 100% of users.

You can either turn your SHOW statement into a new `crdb_internal` view (at the very least) or keep it in `demo.go` altogether.
</blockquote></details>

The plan is to make it work for most users by steering them towards "region" and "zone" keywords.  It's been determined at the PM/TL level that we're willing to take the risk that this won't work for some (on-prem) customers, so that we can deliver a better experience for cloud customers.  Of note is that even many of our on-prem customers (like JPMC, for instance) refer to regions and zones.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLquH1eD9Jhne45M4Sw:bp9j3mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

The plan is to make it work for most users by steering them towards "region" and "zone" keywords.  It's been determined at the PM/TL level that we're willing to take the risk that this won't work for some (on-prem) customers, so that we can deliver a better experience for cloud customers.  Of note is that even many of our on-prem customers (like JPMC, for instance) refer to regions and zones.
</blockquote></details>

ok thanks for the explanation.

Another thing I am seeing here is the switch from `gossip_nodes` to `kv_node_status`.  Was this intentional?

Finally, both versions are inadequate for use by a tenant in a multi-tenant setup. Wasn't SHOW REGIONS meant to be supported for use by tenants? This may hint that using a SQL query over the zone table is not the right approach to do this -- and instead there's a need on a new API on the KV server.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLrBYtYDUM_bsJBgR34:b-l6pgvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ok thanks for the explanation.

Another thing I am seeing here is the switch from `gossip_nodes` to `kv_node_status`.  Was this intentional?

Finally, both versions are inadequate for use by a tenant in a multi-tenant setup. Wasn't SHOW REGIONS meant to be supported for use by tenants? This may hint that using a SQL query over the zone table is not the right approach to do this -- and instead there's a need on a new API on the KV server.
</blockquote></details>

Yes, the switch from gossip_nodes to kv_node_status was intentional (https://github.com/cockroachdb/cockroach/issues/56332).  The switch is mainly due to the fact that gossip can be unreliable (slow to propagate new node info, nodes remain in there for 24 hours after they've been decommissioned) while the kv_node_status will only show region info from nodes that are currently part of the cluster (have been started at some point and have not yet been decommissioned).

The SHOW REGIONS command will be renamed to SHOW REGIONS FROM CLUSTER in my next PR (I don't want to overload this one too heavily).  The intent here is to show all regions available to the cluster.  There is the thought of adding SHOW REGIONS FROM TENANT (and ADD REGIONS TO TENANT) but there's still some debate over whether or not that provides enough value to be implemented (https://cockroachlabs.slack.com/archives/C4ERHM60Z/p1604529403277000).  Either way, there's no plan to get to the tenant piece in 21.1.  


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLrEAhPDPL-e4yBDxKj:bchfs8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Yes, the switch from gossip_nodes to kv_node_status was intentional (https://github.com/cockroachdb/cockroach/issues/56332).  The switch is mainly due to the fact that gossip can be unreliable (slow to propagate new node info, nodes remain in there for 24 hours after they've been decommissioned) while the kv_node_status will only show region info from nodes that are currently part of the cluster (have been started at some point and have not yet been decommissioned).

The SHOW REGIONS command will be renamed to SHOW REGIONS FROM CLUSTER in my next PR (I don't want to overload this one too heavily).  The intent here is to show all regions available to the cluster.  There is the thought of adding SHOW REGIONS FROM TENANT (and ADD REGIONS TO TENANT) but there's still some debate over whether or not that provides enough value to be implemented (https://cockroachlabs.slack.com/archives/C4ERHM60Z/p1604529403277000).  Either way, there's no plan to get to the tenant piece in 21.1.  
</blockquote></details>

If your PR is intended to address #56332 then this should be indicated by "fixes #56332" in the PR description. This ensures that github relates them to each other and closes the one automatically when the other is merged.

Beyond this it looks to me that this commit really has two different changes embedded: one in the "show regions" logic. and another in the demo code. This suggests to me that they should be split into different commits.

Also, I am noticng that the name of the columsn returned by this SHOW statement have changed. Given this is a user-visible feature, I would recommend a release note
(if this had been already publicized in v20.2, you'd also need a bit more verbosity and care, but luckily it wasn't so you're in the clear)


<!-- Sent from Reviewable.io -->

ajstorm committed:
sql: Change SHOW REGIONS to pull from kv_node_status table

Previously the SHOW REGIONS statement pulled data from the
crdb_internal.gossip_nodes table.  The problem with this table however,
is that it isn't a consistent and reliable source of information.  For
example, nodes remain in the gossip table for 24 hours after they've
been shutdown and decommissioned.  Additionally, nodes may not be
present in the gossip_nodes table immediately after they've been added
to the cluster.  To fix these issues, SHOW REGIONS will now pull data
from the kv_node_status table, to which nodes are added when they are
first started, and removed when they are decommissioned.

This commit fixes #56332.

Release note (sql change): SHOW REGIONS has changed the column name for
availability zones to "zones" from "availability_zones".  This was
agreed upon with the TLs and PMs for multi-region simplification.

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)

___
*[pkg/sql/delegate/show_regions.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-MLTgFQu3uwd_zhZ3pB9:-MLreKpQBaV_fkNifmC_:b7ndxo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1502db2cdadac99588e167cafa155c9ebaaa846/pkg/sql/delegate/show_regions.go#L21)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If your PR is intended to address #56332 then this should be indicated by "fixes #56332" in the PR description. This ensures that github relates them to each other and closes the one automatically when the other is merged.

Beyond this it looks to me that this commit really has two different changes embedded: one in the "show regions" logic. and another in the demo code. This suggests to me that they should be split into different commits.

Also, I am noticng that the name of the columsn returned by this SHOW statement have changed. Given this is a user-visible feature, I would recommend a release note
(if this had been already publicized in v20.2, you'd also need a bit more verbosity and care, but luckily it wasn't so you're in the clear)
</blockquote></details>

Done all of the above.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
LGTM but the release note in the first commit is ungrammatical:

>  Release note (cli change): cockroach demo allows for add nodes in a
--geo-partitioned-replicas configuration

Maybe make that:

```
Release note (cli change): `cockroach demo` now allows the user to add nodes
using the `\demo` client-side command, even in combination with 
`--geo-partitioned-replicas`.
```

> Reviewed 1 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-:-MLs4m2F1Wggmszv6JY4:bshi8ef)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)



<!-- Sent from Reviewable.io -->

knz commented:
also don't forget to update the PR description at the top from the commit messages.

ajstorm reviewed (commented):
Done.  Thanks again for the review @knz !

@otan anything else on your end?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56344#-:-MLsHQpl3gg3B6LzPiQa:beaux81)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
Nah go for it

On Thu, 12 Nov 2020, 4:11 am ajstorm, <notifications@github.com> wrote:

> *@ajstorm* commented on this pull request.
>
> Done. Thanks again for the review @knz <https://github.com/knz> !
>
> @otan <https://github.com/otan> anything else on your end?
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/56344#-:-MLsHQpl3gg3B6LzPiQa:beaux81>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on @dt
> <https://github.com/dt>, @knz <https://github.com/knz>, and @otan
> <https://github.com/otan>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/56344#pullrequestreview-528371276>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQZRKBFINNEXOJRN7DLSPLAVJANCNFSM4TL6SB5A>
> .
>

ajstorm commented:
bors r=orange,rnz
